torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  // After:  This TV -> [Set Op] -> New CA TV -> [Use Op] -> Next TV
+
+  // Expr* consumer_uses =
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
+  for (auto expr : uses_expr) {
+    createExprProducer(expr, this, consumer);
+  }
+
+  // Expr* consumer_origin =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // Before: This TV -> Next TV
+  // After:  This TV -> New TV (After) -> Next TV
+  if (this->hasComputeAt()) {
+    TransformReplay::replayCasP(consumer, producer, -1);
+
+    auto rel_ca_pos = this->getRelativeComputeAtAxis();
+    auto this_ca_pos = this->getThisComputeAtAxis();
+    auto this_ca_view = this->getComputeAtView();
+
+    this->computeAt(consumer, this_ca_pos);
+    consumer->setComputeAt(this_ca_view, rel_ca_pos);
+  } else {
+    // Check users of this TV for computeAt for cache_after on inputs
+    std::unordered_set<Expr*> consumer_uses =
+        this->fusion()->unordered_uses(consumer);
+    for (auto expr : consumer_uses) {
+      auto expr_outputs = expr->outputs();
+      for (Val* v : expr_outputs) {
+        if (v->getValType().value() == ValType::TensorView) {
+          TensorView* output = dynamic_cast<TensorView*>(v);
+          if (output->hasComputeAt()) {
+            TransformReplay::replayPasC(consumer, output, -1);
+            auto output_ca_pos = output->getThisComputeAtAxis();
+            consumer->setComputeAt(output, output_ca_pos);
+          }
+        }
+      }
+    }
+  }
+
+  return consumer;
+}
+
+// Create New Expr given consumer - [output of the expression]
+struct CreateExprConsumer : public OptInDispatch {
+ public:
+  static void create(Expr* expr, TensorView* consumer) {
+    CreateExprConsumer cec(consumer);
+    cec.handle(expr);
+  }
+
+ private:
+  explicit CreateExprConsumer(TensorView* consumer) : consumer_(consumer) {}
+
+  void handle(Expr* expr) final {
+    OptInDispatch::handle(expr);
+  }
+
+  void handle(UnaryOp* unary_expr) final {
+    new UnaryOp(unary_expr->getUnaryOpType(), consumer_, unary_expr->in());
+  }
+
+  void handle(BinaryOp* binary_expr) final {
+    new BinaryOp(
+        binary_expr->getBinaryOpType(),
+        consumer_,
+        binary_expr->lhs(),
+        binary_expr->rhs());
+  }
+
+  void handle(TernaryOp* ternary_expr) final {
+    new TernaryOp(
+        ternary_expr->getTernaryOpType(),
+        consumer_,
+        ternary_expr->in1(),
+        ternary_expr->in2(),
+        ternary_expr->in3());
+  }
+
+  void handle(ReductionOp* reduction_expr) final {
+    new ReductionOp(
+        reduction_expr->getReductionOpType(),
+        reduction_expr->init(),
+        consumer_,
+        reduction_expr->in());
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(consumer_, broadcast_expr->in());
+  }
+
+ private:
+  TensorView* consumer_;
+};
+
+// Create New Expr given producer - [an input for the expression]
+struct CreateExprProducer : public OptInDispatch {
```

Comment:
This one should also be in an anonymous namespace?

Reply:
Done